geo/src/algorithm/voronoi_diagram.rs

```diff
@@ -0,0 +1,970 @@
+    // The common point is the end co-ordinate of both lines
+    fn from_midpoints<T: GeoFloat>(line_a: &Line<T>, line_b: &Line<T>) -> GuidingDirection {
+        let dx_a = line_a.dx().is_sign_positive();
+        let dy_a = line_a.dy().is_sign_positive();
+        let dx_b = line_b.dx().is_sign_positive();
+        let dy_b = line_b.dy().is_sign_positive();
+
+        // In a triangle a midpoint must either be
+        // to the left or right of the other midpoints, or
+        // above of below the other midpoints
+        if (dx_a == dx_b) && dx_a {
+            GuidingDirection::Right
+        } else if (dx_a == dx_b) && !dx_a {
+            GuidingDirection::Left
+        } else if (dy_a == dy_b) && dy_a {
+            GuidingDirection::Up
+        } else {
+            GuidingDirection::Down
+        }
+    }
+}
+
+fn define_edge_to_infinity<T: GeoFloat>(
+    triangle: &DelaunayTriangle<T>,
+    circumcenter: &Coord<T>,
+    shared_edges: &mut Vec<Line<T>>,
+    clipping_mask: &Rect<T>,
+) -> Result<Option<Line<T>>>
+where
+    f64: From<T>,
+{
+    let two = T::from(2.).ok_or(VoronoiDiagramError::CannotConvertBetweenGeoGenerics)?;
+
+    let tri: Triangle<T> = triangle.clone().into();
+    let midpoints: Vec<_> = tri
+        .to_lines()
+        .iter()
+        .map(|edge| edge.start + coord! {x: edge.dx() / two, y: edge.dy() / two})
+        .collect();
+    for (edge, midpoint) in tri.to_lines().iter().zip(&midpoints) {
+        let flipped_edge = Line::new(edge.end, edge.start);
+        if shared_edges.contains(edge) || shared_edges.contains(&flipped_edge) {
+            continue;
+        }
+
+        // Get the line that passes from the circumcenter and is perpendicular to
+        // the edge without a 3rd vertex
+        let line: Line<T> = get_perpendicular_line(edge)?;
+        let slope = line.slope();
+
+        // Get the width and height of the clipping mask to ensure intersection with
+        // the lines of infinity. 2 x the width or height should be sufficient.
+        // Add the values to each other to prevent the need for another trait
+        let width_factor = clipping_mask.width() + clipping_mask.width();
+        let height_factor = clipping_mask.height() + clipping_mask.height();
+
+        // We want to move away from the other two edges of the triangle towards infinity.
+        // We will compare the three midpoints of the triangle to determine the correct direction.
+        let rel_to_midpoint: Vec<_> = midpoints
+            .iter()
+            .filter(|x| *x != midpoint)
+            .map(|x| Line::new(*x, *midpoint))
+            .collect();
+        let guiding_direction =
+            GuidingDirection::from_midpoints(&rel_to_midpoint[0], &rel_to_midpoint[1]);
+
+        // It is a vertical line so we need to ensure it moves up or down correctly
+        let end = if slope.is_infinite() {
+            match guiding_direction {
+                GuidingDirection::Left => todo!(),
```

Comment:
Do we need to implement the left and right here?

Reply:
Maybe as errors?

Reply:
left and right here shouldn't be possible.  I was supposed to add an error.  I'll do this now